### ISSUE TITLE
Event System

### ISSUE THREAD

**Issue by da3dsoul** [japanesemediamanager/ShokoServer#591]
The concept is as such, I won't try to define what events are: [Wikipedia](https://en.wikipedia.org/wiki/Event-driven_programming)
This will aid in webui development as well, especially for first time setup, and to make sure proper responses are sent between API start and server loading.
The various events will have attached info based on what is relevant. This will be not only a core system change, but events may be subscribed to and triggered via API, allowing more feature rich addons.
## Proposed events:
(these may need to be split up and organized, they are currently in order in which I thought of them)
- File Added (triggered on the adding of any file, may be a multiple, or even unrecognized)
- Episode Added (only triggered when a new episode is added, which only occurs when a file is added, recognized, and not a multiple)
- Series Added
- Group Added
- Server Preinit (triggered to initialize things)
- Server Starting (triggered after initializing all core systems, for addons or minor subsystems that need core to be ready)
- Server Started (After everything is inited, for post init processing or telling addons the server is done loading)
- Import Run
- the rest of the current buttons in the ui...
- Server Stopping (tell things to shut down, mainly for addons)
- User voted (It could be useful idk)
- Episode watched (will have info for whether it was a manual mark as watched)

@japanesemediamanager/developers Feel free to add more to this list via issue editing.

**Comment by da3dsoul** [comment]
@Cazzar you will want to add/edit this I imagine. Your plex stuff could benefit from such a system.

**Comment by Cazzar** [comment]
Considering the core implementation of events in C#, this would be a good addition.
This would also lead us towards the ability to have a simple plugin system.

**Comment by ElementalCrisis** [comment]
Still on the docket @da3dsoul?

**Comment by da3dsoul** [comment]
Let's keep this because it has a list of events, and only some of them are in atm

**Comment by revam** [comment]
i edited the OP to include pre/post import events (since not having a pre-start even in jf is killing me every time i want to add some logic to run before an import in there)

**Comment by da3dsoul** [comment]
How do those work?

**Comment by revam** [comment]
basically you run the events/hooks eefore the import starts, then wait for them to complete, and then you run the events/hooks after the import. if a pre-start event/hook throws then the import is cancelled as well. but the import is already done when the post-omplete events/hooks run so it won't affect the result, but if it throws rhen it will still show up in the logs.

**Comment by Cazzar** [comment]
> i edited the OP to include pre/post import events (since not having a pre-start even in jf is killing me every time i want to add some logic to run before an import in there)

That seems a bit too much, since events in C# are sync we need to consider where we add events, especially in the case of people adding stuff.

We probably need to have a process of actually discussing new events, and what use cases there will be, as I currently see little to no use case for a pre **and** post import hook, we can use one and handle it that way.

**Comment by bigretromike** [comment]
MQTT Broker ?

**Comment by Cazzar** [comment]
> MQTT Broker ?

That's going to be scope for your own plugin IMO.

**Comment by bigretromike** [comment]
Just pointing out that event system like MQTT or rabbitmq are out there if that helps/or someone don't know them.

**Comment by Cazzar** [comment]
> Just pointing out that event system like MQTT or rabbitmq are out there if that helps/or someone don't know them.

Yes, but those require running an external program for event handling. Same with using things like Redis pubsub, etc
Shoko while it's a client-server architecture, running extra programs like that is outside of scope and not an option for an event system.

**Comment by bigretromike** [comment]
> Shoko while it's a client-server architecture, running extra programs like that is outside of scope and not an option for an event system.

I would argue with that one because you could use your current MQTT infrastructure and offload shoko or even integrating it other services :-) But that's just a theory.  But thank for understanding my point.

**Comment by da3dsoul** [comment]
I don't think most people have a message queue infrastructure. I don't

**Comment by bigretromike** [comment]
> I don't think most people have a message queue infrastructure. I don't

Most HA users which is getting bigger and bigger would be the best example of people less techy yet still using MQTT. Most ziggbe integrations use mqtt.

**Comment by da3dsoul** [comment]
Ah I see. I'm not involved in the smart home crowd, but that makes a lot of sense
